### ISSUE TITLE
Accelerate get_shuffle_matrix_a_row_indices (Python for-loop is a runtime bottleneck)

### ISSUE THREAD

**Issue by yzh119** [flashinfer-ai/flashinfer#2934]
## Summary

`get_shuffle_matrix_a_row_indices` in [`flashinfer/utils.py` (L817-L854)](https://github.com/flashinfer-ai/flashinfer/blob/9c5edf598c49c18a5a17068b9492ab8e79ced262/flashinfer/utils.py#L817-L854) uses a Python `for`-loop over `M` rows to build the shuffle permutation indices. This was originally written for unit tests, but it is now called at runtime in hot paths:

- **`flashinfer/fused_moe/core.py`** — called during FP4 MoE execution (lines 109, 136)
- **`flashinfer/quantization/fp4_quantization.py`** — called during FP4 quantization (line 888)

For large `M`, the Python loop becomes a significant performance bottleneck, as reported by vLLM developers.

## Current Implementation

```python
def get_shuffle_matrix_a_row_indices(
    input_tensor: torch.Tensor, epilogue_tile_m: int
) -> torch.Tensor:
    M, K = input_tensor.shape
    shuffle_block_size = get_shuffle_block_size(epilogue_tile_m)
    row_map = srcToDstBlk16RowMap if shuffle_block_size == 16 else srcToDstBlk32RowMap

    row_indices = torch.empty(M, dtype=torch.long)
    for old_row in range(M):                          # <-- Python for-loop over M
        block_idx = old_row // shuffle_block_size
        row_in_block = old_row % shuffle_block_size
        mapped_row_in_block = row_map[row_in_block]
        new_row = block_idx * shuffle_block_size + mapped_row_in_block
        row_indices[new_row] = old_row
    return row_indices
```

## Proposed Fix

Replace the Python for-loop with a vectorized implementation. Options include:

1. **Pure PyTorch vectorization** — use `torch.arange`, `//`, `%`, and tensor indexing to compute the permutation without any loop (simplest, no extra dependency)
2. **Triton kernel** — write a small Triton kernel for the index computation
3. **CUDA kernel** — add a small dedicated kernel

Option 1 (vectorized PyTorch) is likely sufficient and the simplest approach:

```python
def get_shuffle_matrix_a_row_indices(input_tensor, epilogue_tile_m):
    M, K = input_tensor.shape
    shuffle_block_size = get_shuffle_block_size(epilogue_tile_m)
    row_map = srcToDstBlk16RowMap if shuffle_block_size == 16 else srcToDstBlk32RowMap
    row_map_t = torch.tensor(row_map, dtype=torch.long)

    old_rows = torch.arange(M, dtype=torch.long)
    block_idx = old_rows // shuffle_block_size
    row_in_block = old_rows % shuffle_block_size
    mapped_row_in_block = row_map_t[row_in_block]
    new_rows = block_idx * shuffle_block_size + mapped_row_in_block

    row_indices = torch.empty(M, dtype=torch.long)
    row_indices[new_rows] = old_rows
    return row_indices
```

## Priority

**High** — this is on the critical path for FP4 MoE and FP4 quantization workloads and has been flagged by vLLM developers as a performance regression.

**Comment by yzh119** [comment]
cc @youkaichao

**Comment by aleozlx** [comment]
hi please try this helper function flashinfer.fused_moe.core._maybe_get_cached_w3_w1_permute_indices

usage example

https://github.com/flashinfer-ai/flashinfer/blob/main/tests/moe/test_trtllm_gen_fused_moe.py#L456-L483
